File: src/content/docs/topics/cs/coding-problems/1d-dynamic-programming/091-decode-ways-approach3.go
package main

import (
	"fmt"
	"strconv"
)

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func numDecodings(s string) int {
	if len(s) == 0 || s[0] == '0' {
		return 0
	}
	prev2, prev1 := 1, 1
	for i := 1; i < len(s); i++ {
		cur := 0
		if s[i] != '0' {
			cur += prev1
		}
		two, _ := strconv.Atoi(s[i-1 : i+1])
		if two >= 10 && two <= 26 {
			cur += prev2
		}
		prev2, prev1 = prev1, cur
	}
	return prev1
}

func runTests() {
	assert(numDecodings("12") == 2)
	assert(numDecodings("226") == 3)
	assert(numDecodings("06") == 0)
	assert(numDecodings("0") == 0)
	assert(numDecodings("1") == 1)
	assert(numDecodings("11106") == 2)
	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/1d-dynamic-programming/091-decode-ways-approach3.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

final class Solution {
    func numDecodings(_ s: String) -> Int { let c = Array(s); if c[0] == "0" { return 0 }; var twoBack = 1, oneBack = 1; for i in 1..<c.count { var current = c[i] == "0" ? 0 : oneBack; if let pair = Int(String(c[(i - 1)...i])), pair >= 10 && pair <= 26 { current += twoBack }; twoBack = oneBack; oneBack = current }; return oneBack }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:a0e5efb6ce2e2dcdc3cecaf590953827ce3f73f2a78022e9b60b3cfe616950cd
    expectEqual(Solution().numDecodings("12"), 2, "two-options")
    expectEqual(Solution().numDecodings("226"), 3, "three-options")
    expectEqual(Solution().numDecodings("06"), 0, "leading-zero")
    expectEqual(Solution().numDecodings("2101"), 1, "internal-zero")
    expectEqual(Solution().numDecodings("1"), 1, "single-digit")
    // EXCLUDED_VECTOR empty-string: [""] | The published input contains at least one digit.
    // EXCLUDED_VECTOR non-digit: ["1a"] | The input contains decimal digits only.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
